### PR TITLE
Correct wav2vec2-bert inputs_to_logits_ratio

### PR DESCRIPTION
**Pull request by ylacombe**
# What does this PR do?

`chunk_length_s` parameter doesn’t yet work for w2v2-bert, when using the ASR pipeline.

```python
from transformers import pipeline

pipe = pipeline("automatic-speech-recognition", model="ylacombe/wav2vec2-bert-CV16-en-libri-cv", device=0)
pipe("https://huggingface.co/datasets/Narsil/asr_dummy/resolve/main/1.flac", chunk_length_s=20.0, batch_size=16)
```

Turns out the `inputs_to_logits_ratio` wasn't working properly. I'm not totally sure that I've chosen the right input ratio so I'd like @Narsil's opinion on this.

The original W2V2 uses a [learned feature extractor](https://github.com/huggingface/transformers/blob/abbffc4525566a48a9733639797c812301218b83/src/transformers/models/wav2vec2_conformer/modeling_wav2vec2_conformer.py#L486) to downsample the raw audio waveform to a more reasonable dimension, using a series of convolution layer.
However, w2v2-bert does that feature reduction thanks to a mel-spectrogram transformation;

If I understood `inputs_to_logits_ratio`, it only computes the hidden size after this feature reduction, which in the case of w2v2-bert doesn't exist. So I've put `inputs_to_logits_ratio=1`. I could very be wrong on this, WDYT @Narsil ?

cc @sanchit-gandhi

### CODE REVIEW

**Comment by Narsil** [comment]
inputs_to_logit_ratio is:

-> Inputs (raw audio) -> [32000; u32] (4s audio at 8kHz for instance)
-> Logits [2000; u32] (Same 4s audio, just the convolutions has a reduction factor of 16 for instance, I don't remember the actual w2v2 ratio).


Usually mel spectrogram also has a downsampling factor, but I don't remember on the top of my head the name of the controlling factor. Mel Spectro is basically some convolutions on the raw audio too. They just happen to be FFT.

**Comment by ylacombe** [comment]
Thanks @Narsil, the ratio is wrong then, I'll correct it

**Comment by ylacombe** [comment]
For reference, I've double-checked the ratio.

W2V2-Bert uses [M4T feature extractor](https://github.com/huggingface/transformers/blob/0e75aeefaf4beeb7a5bb6a1f05b83ab99e045a24/src/transformers/models/seamless_m4t/feature_extraction_seamless_m4t.py#L38), which reproduces kaldi mel-spectrogram computation.

To compute the ratio:
1. Get ratio from mel-spectrogram. It can be found [here](https://github.com/pytorch/audio/blob/02586da797dbfa201721d2080e2171805202f72c/src/torchaudio/compliance/kaldi.py#L67). It is basically `1 + (waveform_length - frame_length) // hop_length`.
2. Divide by [`stride`](https://github.com/huggingface/transformers/blob/0e75aeefaf4beeb7a5bb6a1f05b83ab99e045a24/src/transformers/models/seamless_m4t/feature_extraction_seamless_m4t.py#L292-L294).

Here `frame_length=400`, `hop_length=160` and `stride=2`. Exact computation gives `logits_length= waveform_length / 320 - 0.75`.

I'll round the ratio to `320` (`hop_length*ratio`). We can get an approximation of the ratio thanks to [`config.feature_projection_input_dim`](https://github.com/huggingface/transformers/blob/0e75aeefaf4beeb7a5bb6a1f05b83ab99e045a24/src/transformers/models/wav2vec2_bert/configuration_wav2vec2_bert.py#L57C9-L57C37) which is computed like this `(hop_length//2) * ratio`.

File: src/transformers/models/wav2vec2_bert/configuration_wav2vec2_bert.py [hunk 1]
```diff
@@ -311,4 +309,7 @@ def __init__(
 
     @property
     def inputs_to_logits_ratio(self):
-        return functools.reduce(operator.mul, self.conv_stride, 1)
+        ratio = self.feature_projection_input_dim * 2
+        if self.add_adapter:
+            ratio = ratio * (self.adapter_stride**self.num_adapter_layers)
+        return ratio
```
Comment (ylacombe) [inline src/transformers/models/wav2vec2_bert/configuration_wav2vec2_bert.py:315]: For reference, I've double-checked the ratio.


W2V2-Bert uses [M4T feature extractor](https://github.com/huggingface/transformers/blob/0e75aeefaf4beeb7a5bb6a1f05b83ab99e045a24/src/transformers/models/seamless_m4t/feature_extraction_seamless_m4t.py#L38), which reproduces kaldi mel-spectrogram computation.
To compute the ratio:
1. Get ratio from mel-spectrogram. It can be found [here](https://github.com/pytorch/audio/blob/02586da797dbfa201721d2080e2171805202f72c/src/torchaudio/compliance/kaldi.py#L67). It is basically `1 + (waveform_length - frame_length) // hop_length`.
2. Divide by [`stride`](https://github.com/huggingface/transformers/blob/0e75aeefaf4beeb7a5bb6a1f05b83ab99e045a24/src/transformers/models/seamless_m4t/feature_extraction_seamless_m4t.py#L292-L294).
3. Multiply by the model downsample factor -> it appears if an adapter block has been added on top of the model

Here, for the two first steps, `frame_length=400`, `hop_length=160` and `stride=2`. Exact computation gives `logits_length= waveform_length / 320 - 0.75`.
I'll round the ratio to `320` (`hop_length*ratio`). We can get an approximation of the ratio thanks to [`config.feature_projection_input_dim`](https://github.com/huggingface/transformers/blob/0e75aeefaf4beeb7a5bb6a1f05b83ab99e045a24/src/transformers/models/wav2vec2_bert/configuration_wav2vec2_bert.py#L57C9-L57C37) which is computed like this `(hop_length//2) * ratio`.

So the ratio of the two first steps should be `feature_projection_input_dim*2`.

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 2]
```diff
@@ -73,13 +73,6 @@ def chunk_iter(inputs, feature_extractor, chunk_len, stride_left, stride_right,
 
         chunk_len = chunk.shape[0]
         stride = (chunk_len, _stride_left, _stride_right)
-        if "input_features" in processed:
-            processed_len = processed["input_features"].shape[-1]
-        elif "input_values" in processed:
-            processed_len = processed["input_values"].shape[-1]
```
Comment (ylacombe) [inline src/transformers/pipelines/automatic_speech_recognition.py:79]: In most cases, this operation does nothing. 

Prior feature extractors of CTC models output the raw audio waveform, so before W2V2-BERT,  `processed_len` always equaled `chunk_len`.

The operation introduces errors for feature extractors that are using spectrograms-related computations because rescaling is done twice, as explained in the following comment.

File: src/transformers/pipelines/automatic_speech_recognition.py [hunk 3]
```diff
@@ -436,10 +429,8 @@ def preprocess(self, inputs, chunk_length_s=0, stride_length_s=None):
             if chunk_len < stride_left + stride_right:
                 raise ValueError("Chunk length must be superior to stride length")
 
-            rescale = self.type != "seq2seq_whisper"
-            # make sure that
             for item in chunk_iter(
-                inputs, self.feature_extractor, chunk_len, stride_left, stride_right, rescale, self.torch_dtype
```
Comment (ylacombe) [inline src/transformers/pipelines/automatic_speech_recognition.py:442]: There is no need to rescale in `chunk_iter`.

1. For `type=seq2seq_whisper`, it was already doing nothing.
2. For `type=seq2seq`, stride is not used (see [here](https://github.com/huggingface/transformers/blob/80d50076c8d28a5f872066c825f9eb14eb737fbf/src/transformers/pipelines/automatic_speech_recognition.py#L552-L567))
3. For `type=ctc`, rescaling the stride is already done once [here](https://github.com/huggingface/transformers/blob/80d50076c8d28a5f872066c825f9eb14eb737fbf/src/transformers/pipelines/automatic_speech_recognition.py#L536-L539).

This wasn't spotted before because the rescaling ratio for ctc models was always 1 before, as pointed out in the previous comment.

**Review by ylacombe — COMMENTED** [review]
At the end of the day, I've recomputed the `inputs_to_logits_ratio`.

I've also cleaned a bit the ASR pipeline code, removing code that was not properly used.

cc @sanchit-gandhi @Narsil and @ArthurZucker

File: tests/pipelines/test_pipelines_automatic_speech_recognition.py [hunk 4]
```diff
@@ -1334,22 +1334,29 @@ def test_chunking_with_lm(self):
     def test_chunk_iterator(self):
         feature_extractor = AutoFeatureExtractor.from_pretrained("facebook/wav2vec2-base-960h")
         inputs = torch.arange(100).long()
-        ratio = 1
-        outs = list(chunk_iter(inputs, feature_extractor, 100, 0, 0, ratio))
+        outs = list(
```
Comment (ArthurZucker) [inline tests/pipelines/test_pipelines_automatic_speech_recognition.py:1337]: should fit in one line

**Review by ArthurZucker — APPROVED** [review]
Makes sense. Would just make sure that slow test for pipeline passe + modeling test 🤗 Nice refactoring!

File: src/transformers/models/wav2vec2_bert/configuration_wav2vec2_bert.py (hunk 1 above)
Comment (sanchit-gandhi) [reply src/transformers/models/wav2vec2_bert/configuration_wav2vec2_bert.py:315]: Great! Thanks for the clear explanation!

**Review by sanchit-gandhi — APPROVED** [review]
Clean PR - thanks for fixing @ylacombe! Good for me to merge if the slow tests for the ASR pipeline pass as before

**Comment by ylacombe** [comment]
Thanks for the review! All slow tests for the ASR pipeline pass! BTW, I'll make a subsequent PR to correct the fact that the long transcription script for s2s models (beside Whisper) is never called

**Comment by ArthurZucker** [comment]
Very good refactoring 😉
